### ISSUE TITLE
Project name of maximum length (255) breaks in spectacular ways

### ISSUE THREAD

**Issue by tianon** [goharbor/harbor#15067]
**Expected behavior and actual behavior:**
When creating a project, the length limit is 255 characters.  However, if a project of this length is created, it cannot be deleted and it cannot accept images (either pushed or replicated).

Even attempting to push them is complicated because most tools mimic (or directly use) the parsing implemented in the Distribution project, which sets the limit for the full `registry-hostname.example.com/foo/bar` portion to 255 characters: https://github.com/distribution/distribution/blob/411d6bcfd2580d7ebe6e346359fa16aceec109d5/reference/reference.go#L37

```console
$ docker tag hello-world e.e/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee
$ docker tag hello-world e.e/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee
Error parsing reference: "e.e/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee" is not a valid repository/tag: repository name must not be more than 255 characters
```

(The first of these is exactly 255, the second adds one more `e` to become exactly 256.)

If we then further assume the registry is Harbor, that absolute extreme maximum becomes `e.e/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee/e` which is a hostname of `e.e`, a project name of 249x`e`, and a repository name of `e`.

However, this is also complicated by two other factors:

- the registry backend will sometimes add a prefix to this (and then validating the result using the code linked above), further limiting the total possible project maximum
- when _deleting_ a project, the name is updated to `name + "#" + project_id`, but the database field only fits 255 characters so the deletion fails completely (and thus needs headroom for an arbitrary length of `project_id` + 1)

My proposal would be to adjust that limit from 255 down to something like 200 to leave 55 characters of headroom for both the registry hostname + repository name and the project ID in the case of deletion, but I think the "create project" dialog should probably also warn that the total limit of these three components together is 255 in practice.

**Steps to reproduce the problem:**
Create a project with 255 characters in the name (although be prepared to hand-edit the database to make it go away again).

**Versions:**
Please specify the versions of following systems.

- harbor version: 2.2.2

**Comment by ywk253100** [comment]
This is a known issue and already be fixed by https://github.com/goharbor/harbor/pull/14912 which will be available in 2.3

**Comment by tianon** [comment]
Apologies, I tried to search for a duplicate issue and couldn't find one 😅

However, that PR only solves the delete half - any thoughts on this limit being too high given where it's used having a much lower effective limit?
